### ISSUE TITLE
PHP7.0 microseconds are round to 0

### ISSUE THREAD

**Issue by nlevee** [census-instrumentation/opencensus-php#159]
The function Span::formatDate construct a `\DateTime` object without microtime (or microtime to '000000') due to a php limitation before version 7.1 ([http://php.net/manual/en/datetime.construct.php#refsect1-datetime.construct-changelog](http://php.net/manual/en/datetime.construct.php#refsect1-datetime.construct-changelog))

**Comment by chingor13** [comment]
Thanks for the report, but it looks like [we're using `microtime()` to build a date string](https://github.com/census-instrumentation/opencensus-php/blob/master/src/Trace/Span.php#L345-L347) for the `DateTime` constructor. Are you using an older version perhaps?

**Comment by bogdandrutu** [comment]
@chingor13 I looked briefly into the documentation of the date http://php.net/manual/en/function.date.php and I saw:

"Microseconds (added in PHP 5.2.2). Note that date() will always generate 000000 since it takes an integer parameter, whereas DateTime::format() does support microseconds if DateTime was created with microseconds."

Also I saw that the timestamp (date argument) is the number of seconds since epoch. Can this be related?

**Comment by bogdandrutu** [comment]
My bad, just learned a bit php and I saw that you actually do a concatenation there.

**Comment by nlevee** [comment]
cf. migration to PHP 7.1 : [http://php.net/manual/fr/migration71.incompatible.php#migration71.incompatible.datetime-microseconds](http://php.net/manual/fr/migration71.incompatible.php#migration71.incompatible.datetime-microseconds)

> DateTime and DateTimeImmutable now properly incorporate microseconds when constructed from the current time, either explicitly or with a relative string (e.g. "first day of next month")

Before 7.1 construct with micrcotime don't work properly, if you do `->format ('u')` the result is equal `0` in php 7.0

**Comment by nlevee** [comment]
Here is my latency in GCE Stackdriver Trace (every line are round to second because of this issue) : 

![image](https://user-images.githubusercontent.com/198266/38388797-99d3f9d6-391c-11e8-92f6-96d4d34b1fff.png)

**Comment by chingor13** [comment]
Hmm, I'm still not sure what's happening in your case.

I added a test in #163 that makes sure that milliseconds are not 0 in a simple span with defaults. I also pushed a sample application to GAE Flex that is using the Stackdriver Trace exporter and I'm getting results in ms.

Can you post the exact versions of the libraries (opencensus/opencensus and google/cloud or google/cloud-trace) and PHP you are using? Do you have a small code sample that replicates this?

**Comment by nlevee** [comment]
I find something wired, on command line it's work : 

```bash
php -r 'echo (new \DateTime("2018-04-06 09:42:14.999273"))->format("U.u");'
# 1523000534.999273
```

But with Apache microsecond are returned = '000000'

**Comment by chingor13** [comment]
@nlevee Thanks for the extra info. I'll try and reproduce this under Apache.

As a side note, we try to guess the root span's start time from the `$_SERVER` array under `REQUEST_TIME_FLOAT` or `REQUEST_TIME` if available. `REQUEST_TIME` doesn't have sub-second resolution, but I don't think this is the case for you as you're providing the `startTime` using `microtime(true)`

**Comment by chingor13** [comment]
I put together [a simple test case](https://github.com/chingor13/php-test-apps/tree/opencensus/oc-apache) that runs php 7.0 (7.0.29) with apache (2.4.10) via the official PHP docker image. I even set the timezone setting in `php.ini` to `Europe/Paris`.

```bash
$ docker build -t opencensus-apache .
$ docker run -d --rm -p8080:80 opencensus-apache
$ curl http://localhost:8080/time.php
object(DateTime)#1 (3) {
  ["date"]=>
  string(26) "2018-04-06 19:55:10.180521"
  ["timezone_type"]=>
  int(3)
  ["timezone"]=>
  string(12) "Europe/Paris"
}
string(17) "1523037310.180521"
```

**Comment by chingor13** [comment]
@nlevee Since results are different between CLI and via Apache, it sounds like there might be a difference between `php.ini` settings. Is it possible to post your `php.ini` settings as well?

**Comment by chingor13** [comment]
Most likely, we'll switch to `DateTime::createFromFormat('U.u', ...)`, but it will still be nice to figure out why this is happening.

**Comment by nlevee** [comment]
I find why ! I set the locale in fr_FR.UTF-8 at the very begining, without this line, i have the microsecond, with this line i do not have :
```
setlocale(LC_ALL, 'fr_FR.UTF-8');
```

My bad for this issue, it's more a bug in php than your library ...

**Comment by chingor13** [comment]
Ah, yes, I can repro it with setting the locale.

However, setting the locale is totally a valid use case that we should be handling. I *believe* the `DateTime` constructor is using some locale based configuration for parsing the date string. Switching to `DateTime::createFromFormat` with `U.u` seems to be the correct solution here (Fixing in #168).

**Comment by bogdandrutu** [comment]
@chingor13 thanks a lot for fixing this issue.

**Comment by chingor13** [comment]
Closed via #168

**Comment by nlevee** [comment]
The ref to the bug in PHP bug tracker : https://bugs.php.net/bug.php?id=67127

It's very old...
